### PR TITLE
BuildInfo: Support more basic types

### PR DESCRIPTION
**Pull request by benjaminfrank**
Support more basic types in BuildInfo.

### CODE REVIEW

File: contrib/buildinfo/src/mill/contrib/BuildInfo.scala [hunk 1]
```diff
@@ -13,17 +13,23 @@ trait BuildInfo extends ScalaModule {
 
   def buildInfoObjectName: String = "BuildInfo"
 
-  def buildInfoMembers: T[Map[String, String]] = T {
-    Map.empty[String, String]
-  }
+  def buildInfoMembers: Map[String, Any] = Map.empty[String, Any]
```
Comment (Ichoran) [inline contrib/buildinfo/src/mill/contrib/BuildInfo.scala]: This seems really type-unsafe.  I really wouldn't want this in unless you wrapped the types in an ADT.

Comment (benjaminfrank) [reply contrib/buildinfo/src/mill/contrib/BuildInfo.scala]: I had an at ADT first but found the usability lacking: to support implicit conversion you had to put the type arguments to your map or you had to manually wrap everything yourself.
While the current approach only supports the basic types and for everything else basically relies on `toString´.
Is there a particular scenario you'd be afraid of that could happen where this leads to an issue?

File: contrib/buildinfo/test/src/mill/contrib/BuildInfoTests.scala [hunk 2]
```diff
@@ -17,30 +17,55 @@ import utest.framework.TestPath
...
+  val doubleVal: Double = 1.79769313486231570e+308d
+  val charVal: Char = 'Æ'
+  val stringVal: String = "2.12.4"
+
+  val packageName = "foo"
+  val objectName = "bar"
   trait BuildInfoModule extends TestUtil.BaseModule with scalalib.ScalaModule with BuildInfo {
     def millSourcePath =  TestUtil.getSrcPathBase() / millOuterCtx.enclosing.split('.')
-    def scalaVersion = scalaVersionString
+    def scalaVersion = "2.12.6"
```
Comment (Ichoran) [inline contrib/buildinfo/test/src/mill/contrib/BuildInfoTests.scala]: Why is this hard-coded?

Comment (benjaminfrank) [reply contrib/buildinfo/test/src/mill/contrib/BuildInfoTests.scala]: it's the build for the test and needs a scalaversion.

**Comment by benjaminfrank** [comment]
@rockjam @lihaoyi any opinion om the typesafety issue here?
only drawback I found when using an ADT was that the Map requires an type annottaion when using implict conversion, e.g., 
``` scala
def buildInfoMembers = 
  Map[String, ADT](
    "intVal" -> 1)
```
or using the ADT directly, e.g.,
``` scala
def buildInfoMembers  = 
  Map.empty(
    "intVal" -> ADTInt(1))
```

**Comment by lihaoyi** [comment]
I don't care much either way, but if you want to avoid the type ascription you can always define your own `BuildInfo(args: (String, ADT)*)` function so people can `def buildInfoMembers = BuildInfo("intVal" -> 1)` and have the implicits kick in correctly

**Comment by benjaminfrank** [comment]
hm didn't think of that. I will give it a try later. thanks

**Comment by lihaoyi** [comment]
closing due to inactivity

**Pull request closed without merging**